Resize direct lvm

chenchun opened the pull request:
Improves device_tool to be able to reload pool && adds docs about how to extend devicemapper thin-pool

thaJeztah commented:
/cc @rhvgoyal ptal

Thanks @chenchun

rhvgoyal commented:
@chenchun  Why are you introducing this new `reload` command. How does it help?

You seem to be changing block size. Can you really change block size of a running thin pool? I thought block size gets fixed at pool creation time and after that can't be changed?

chenchun commented:
@rhvgoyal I'm not changing block size of the pool. I'm increasing pool capacity.
The added docs shows how to increase the capacity if the `data` lv is full and `vg-docker` volume-group has no more capacity.
```
#### direct-lvm mode

1. Extend volume group (VG) `vg-docker`

        $ vgextend vg-docker /dev/sdh
        Volume group "vg-docker" successfully extended

2. Extend the `data` logical volume(LV) `vg-docker/data`

        $ lvextend  -l+100%FREE -n vg-docker/data
        Extending logical volume data to 200 GiB
        Logical volume data successfully resized

3. Reload devicemapper thin-pool with device_tool

        $ ./device_tool reload /dev/vg-docker/data /dev/vg-docker/metadata
```

> Why are you introducing this new reload command. How does it help?

If I only do the above 1 & 2 steps, the num_sectors of devicemapper thin pool still stays the original size and I have to reload the thin pool to make it refresh. 

I'm following the steps mentioned in https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Storage_Administration_Guide/ch37s03.html#procedure-multipathd-afterresize-pre_RHEL5u3 to fresh num_sectors. We can refresh it by using dmsetup command as the document states, but I think it's better to improve `device_tool` to make it simple.

rhvgoyal commented:
@chenchun Why don't you use lvm to create lvm thin pool and pass it to docker using `dm.thinpooldev`. you can look at `man lvmthin` to figure out how to create lvm thin pool or use `docker-storage-setup` to set one for you.

https://github.com/projectatomic/docker-storage-setup

Point being, device-tool is not a proper tool to manage thin pool. My understanding is that this is more a debugging tool. lvm can already do whatever you are trying to do with device_tool. What's the point of building lvm capabilities in this device_tool. Using lvm to create lvm thin pool and using lvm to manage that pool is easier option then keep on adding more stuff to device_tool.

thaJeztah commented on docs/userguide/storagedriver/device-mapper-driver.md:
```diff
@@ -304,6 +304,35 @@ options were created in the previous steps.
+
+#### compile device_tool
+
+You have to compile contrib/docker-device-tool/device_tool.go [reference](../../../contrib/docker-device-tool/README.md)
```
You need to change this to an absolute URL to https://github.com/....., because the "contrib" directory is not part of the documentation

chenchun commented:
@rhvgoyal https://github.com/projectatomic/docker-storage-setup is really professional. But I did not notice that before and I was following docker official docs https://docs.docker.com/engine/userguide/storagedriver/device-mapper-driver/ to create direct-lvm for production. My problem is that devicemapper pool is full on serveral nodes now. I want to extend the pool on the fly because I'm running some important containers on those nodes and it is painful to stop those containers. And it seems there is not any docs talking about how to extend the pool on the fly.

rhvgoyal commented:
@chenchun So you need a new docker to grow your thin pool. That means you will have to restart docker. And thin pool is reloaded anyway during restart so it will automatically grow. So I am not sure how this tool will help you.

Anyway, why does user need to pass "datadevice" and "metadatadevice" for reload operation. Graphdriver already knows about data device and metadata device and same can be retrieved internally and used?

rhvgoyal commented on contrib/docker-device-tool/README.md:
```diff
@@ -0,0 +1,11 @@
+
+    $ make shell
+    ## inside build container
+    $ go build -tags libdm_no_deferred_remove contrib/docker-device-tool/device_tool.go
```
Why should it compiled with tag -libdm_no_deferred_remove?

chenchun replied:
Because the devicemapper version installed in image does not contain this flag.

```
root@1903dadfb993:/go/src/github.com/docker/docker# go build  contrib/docker-device-tool/device_tool.go
# github.com/docker/docker/pkg/devicemapper
could not determine kind of name for C.dm_task_deferred_remove
root@1903dadfb993:/go/src/github.com/docker/docker# go build -tags libdm_no_deferred_remove contrib/docker-device-tool/device_tool.go
root@1903dadfb993:/go/src/github.com/docker/docker#
```

rhvgoyal replied:
That's specific to certain images. I could build device_tool on fedora host and there it should build just fine without this tag. You might want to specifically mention that if devicemapper version is old and compliation fails, use this tag.

rhvgoyal commented on docs/userguide/storagedriver/device-mapper-driver.md:
```diff
@@ -304,6 +304,35 @@ options were created in the previous steps.
+You have to compile contrib/docker-device-tool/device_tool.go [reference](https://github.com/docker/docker/blob/master/contrib/docker-device-tool/README.md)
+
+#### loop-lvm mode
+
```
There is no lvm involved here. Not sure why are you calling it loop-lvm

chenchun replied:
I'm following the above document `Docker hosts running the devicemapper storage driver default to a configuration mode known as loop-lvm`

rhvgoyal commented:
Why can't you use `dmsetup` to reload your thin pool. Seriously, I am really not in favor of converting device_tool into another tool which will try to do what dmsetup and lvm already do.

rhvgoyal commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2562,3 +2562,35 @@ func NewDeviceSet(root string, doInit bool, options []string, uidMaps, gidMaps [
+
+	// Reload with the new block sizes
+	if err := devicemapper.ReloadPool(devices.getPoolName(), datafile, metadatafile, devices.thinpBlockSize); err != nil {
+		return fmt.Errorf("Unable to reload pool: %s", err)
```
What happens if user is running docker with dm.blocksize=X which is different from default block size. Then this logic is broken.

chenchun commented:
> So you need a new docker to grow your thin pool. That means you will have to restart docker. 

I'm not sure this is true. I can extend pool without restarting docker with the above steps. 

> Anyway, why does user need to pass "datadevice" and "metadatadevice" for reload operation. 

This is kind of possible. I'll have a try.

> Why can't you use dmsetup to reload your thin pool. Seriously, I am really not in favor of converting device_tool into another tool which will try to do what dmsetup and lvm already do.

Because it's a little complicated to use dmsetup directly. You have to do it manually. 
1. dmsetup table thin_pool_device
2. dmsetup suspend thin_pool_device
3. Change the second number of table info (i.e. the disk end sector) to reflect the new number of 512 byte sectors in the disk.
4. dmsetup reload thin_pool_device table_name
5. dmsetup resume thin_pool_device

Besides we can already use `device_tool` to resize loop device backend thin pool. So I think it's better to make the above dmsetup steps implemented in `device_tool`.
Anyway, if `device_tool` is not the recommended way, I can change the docs to use dmsetup directly.

chenchun committed:
Add docs about how to extend devicemapper thin pool

Signed-off-by: Chun Chen <ramichen@tencent.com>

chenchun commented:
@rhvgoyal @thaJeztah 
I removed the code of reloading pool with device_tool and updated docs to only use `dmsetup` to reload direct-lvm thin pool. 
I think it's helpful to add these docs because the official docs encourages people to use direct-lvm mode with docker created pool in production and I believe not only me have the demand of extending thin pool on the fly. And someone may accidentally using loop device on production, so I also add docs about how to extend loop device backend thin pool.

rhvgoyal commented on docs/userguide/storagedriver/device-mapper-driver.md:
```diff
@@ -304,6 +304,146 @@ options were created in the previous steps.
+Alternatively you can simply use device_tool to resize the loop-lvm thin pool, all you have to do is executing the following single command.
+
+        $ ./device_tool resize 200GB
+
```
Is this still valid? I thought we are not modifying device_too for this.

rhvgoyal commented:
In general I like the idea of documenting that how to extend thin pool once it gets full. Please refresh your patch and then I will review it further.

thaJeztah commented:
/cc @moxiegirl, because I think you also were working in changes in this area

thaJeztah commented:
ping @moxiegirl could you have a look at this one?

moxiegirl commented on docs/userguide/storagedriver/device-mapper-driver.md:
```diff
@@ -304,6 +304,146 @@ options were created in the previous steps.
  The last two lines also confirm that the correct devices are being used for
 the `Data file` and the `Metadata file`.
 
+### Extend devicemapper thin-pool
```
@chenchun Thank you for the contribution.  There are some formatting issues in here and other issues.  Please move this section to starting at line 366 *after* the basic configuration information. Then, change the `####` to `###` and `###` to `##`. I've provide a gist with these changes (except for the move)

---> https://gist.github.com/moxiegirl/7edfe43486944cd487d3da66d9d0b400

## Increase capacity on a running device

You can increase the capacity of the pool on a running thin-pool device. This is
useful if the data's logical volume is full and the volume group is at full
capacity.

### For a loop-lvm configuration

In this scenario, the thin pool is configured to use `loop-lvm` mode. To show the specifics of the existing configuration use `docker info`:

```bash
$ sudo docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.11.0-rc2
Storage Driver: devicemapper
 Pool Name: docker-8:1-123141-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.202 GB
 Data Space Total: 107.4 GB
 Data Space Available: 4.506 GB
 Metadata Space Used: 1.729 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.90 (2014-09-01)
Logging Driver: json-file
[...]
```

The `Data Space` values show that the pool is 100GB total. This example extends the pool to 200GB.

1. List the sizes of the devices.

	```bash
	$ sudo ls -al /var/lib/docker/devicemapper/devicemapper/
	total 1175492
	drwx------ 2 root root         4096 Mar 29 02:45 .
	drwx------ 5 root root         4096 Mar 29 02:48 ..
	-rw------- 1 root root 107374182400 Mar 30 05:22 data
	-rw------- 1 root root   2147483648 Mar 31 11:17 metadata
	```

2. Truncate `data` file to the size of the `metadata` file (approximage 200GB).

	```bash
	$ sudo truncate -s 214748364800 /var/lib/docker/devicemapper/devicemapper/data
	```

3. Verify the file size changed.

	```bash
	$ sudo ls -al /var/lib/docker/devicemapper/devicemapper/
	total 1175492
	drwx------ 2 root root         4096 Mar 29 02:45 .
	drwx------ 5 root root         4096 Mar 29 02:48 ..
	-rw------- 1 root root 214748364800 Mar 31 11:20 data
	-rw------- 1 root root   2147483648 Mar 31 11:17 metadata
	```

4. Reload data loop device

	```bash
	$ sudo blockdev --getsize64 /dev/loop0
	107374182400
	$ sudo losetup -c /dev/loop0
	$ sudo blockdev --getsize64 /dev/loop0
	214748364800
	```

5. Reload devicemapper thin pool.

	a. Get the pool name first.

		$ sudo dmsetup status docker-8:1-123141-pool: 0 209715200 thin-pool 91
		422/524288 18338/1638400 - rw discard_passdown queue_if_no_space -

		The name is the string before the colon.

 	b. Dump the device mapper table first.

		$ sudo dmsetup table docker-8:1-123141-pool
		0 209715200 thin-pool 7:1 7:0 128 32768 1 skip_block_zeroing

	c. Calculate the real total sectors of the thin pool now.

		Change the second number of the table info (i.e. the disk end sector) to reflect the new number of 512 byte sectors in the disk. For example, as the new loop size is 200GB, change the second number to 419430400.

	d. Reload the thin pool with the new sector number

		$ sudo dmsetup suspend docker-8:1-123141-pool && sudo dmsetup reload docker-8:1-123141-pool --table '0 419430400 thin-pool 7:1 7:0 128 32768 1 skip_block_zeroing' && sudo dmsetup resume docker-8:1-123141-pool

#### The device_tool

The Docker's projects `contrib` directory contains not part of the core
distribution. These tools that are often useful but can also be out-of-date. <a
href="https://goo.gl/wNfDTi">In this directory, is the `device_tool.go`</a>
which you can also resize the loop-lvm thin pool.

To use the tool, compile it first. Then, do the following to resize the pool:

```bash
$ ./device_tool resize 200GB
```

### For a direct-lvm mode configuration

In this example, you extend the capacity of a running device that uses the
`direct-lvm` configuration.  This example assumes you are using the `/dev/sdh1`
disk partition.

1. Extend the volume group (VG) `vg-docker`.

	```bash
	$ sudo vgextend vg-docker /dev/sdh1
	Volume group "vg-docker" successfully extended
	```

	Your volume group may use a different name.

2. Extend the `data` logical volume(LV) `vg-docker/data`

	```bash
	$ sudo lvextend  -l+100%FREE -n vg-docker/data
	Extending logical volume data to 200 GiB
	Logical volume data successfully resized
	```

3. Reload devicemapper thin pool.

	a. Get the pool name.

		$ sudo dmsetup status | grep pool
		docker-253:17-1835016-pool: 0 96460800 thin-pool 51593 6270/1048576 701943/753600 - rw no_discard_passdown queue_if_no_space

		The name is the string before the colon.

	b. Dump the device mapper table.

		$ sudo dmsetup table docker-253:17-1835016-pool
		0 96460800 thin-pool 252:0 252:1 128 32768 1 skip_block_zeroing

	c. Calculate the real total sectors of the thin pool now. we can use `blockdev` to get the real size of data lv.

		Change the second number of the table info (i.e. the disk end sector) to
		reflect the new number of 512 byte sectors in the disk. For example, as the
		new data `lv` size is `264132100096` bytes, change the second number to
		`515883008`.

		$ sudo blockdev --getsize64 /dev/vg-docker/data
		264132100096

	d. Then reload the thin pool with the new sector number.
	
		$ sudo dmsetup suspend docker-253:17-1835016-pool && sudo dmsetup reload docker-253:17-1835016-pool --table  '0 515883008 thin-pool 252:0 252:1 128 32768 1 skip_block_zeroing' && sudo dmsetup resume docker-253:17-1835016-pool

chenchun replied:
Thanks, refreshed the patch with your gist and changed a typo and a format issue

thaJeztah commented:
thanks! LGTM

ping @moxiegirl ptal

rhvgoyal commented on docs/userguide/storagedriver/device-mapper-driver.md:
````diff
@@ -397,6 +397,176 @@ there are two key directories. The `/var/lib/docker/devicemapper/mnt` directory
+
+```bash
+$ ./device_tool resize 200GB
+```
````
So this will work even if I am running thin pool with a non-default block size?

chenchun replied:
Good point. It seems device_tool passes the default block size which is 64K. But I tried it just now, it works. I suspect it is the kernel who silently prevents changing the block size.

rhvgoyal commented on docs/userguide/storagedriver/device-mapper-driver.md:
````diff
@@ -397,6 +397,176 @@ there are two key directories. The `/var/lib/docker/devicemapper/mnt` directory
+	```
+
+2. Truncate `data` file to the size of the `metadata` file (approximate 200GB).
+
````
Why truncate data file to the size of metadata file? I thought you want to truncate data file to 200GB?

rhvgoyal commented on docs/userguide/storagedriver/device-mapper-driver.md:
````diff
@@ -397,6 +397,176 @@ there are two key directories. The `/var/lib/docker/devicemapper/mnt` directory
+
+	```bash
+	$ sudo ls -al /var/lib/docker/devicemapper/devicemapper/
+	total 1175492
````
It might be better to give these examples with ls option "-h" to make it human readable.

rhvgoyal commented on docs/userguide/storagedriver/device-mapper-driver.md:
```diff
@@ -397,6 +397,176 @@ there are two key directories. The `/var/lib/docker/devicemapper/mnt` directory
+	c. Calculate the real total sectors of the thin pool now.
+
+		Change the second number of the table info (i.e. the disk end sector) to reflect the new number of 512 byte sectors in the disk. For example, as the new loop size is 200GB, change the second number to 419430400.
+
```
I think second number reflects number of sectors and not the end sector. Look at "man dmsetup" in section TABLE FORMAT.

rhvgoyal commented:
Apart from minor comments, this LGTM. Good documentation. We need this to grow thin pool. 

Again I am wondering how to integrate docker-storage-setup better. There we setup and use lvm thin pool which automatically grows as thin pool gets filled based on certain thresholds. If volume group is about to be full, all one has to do to add storage to volume group and it will extend automatically.

thaJeztah commented:
@rhvgoyal would it be an idea to contribute it to this repository (for inclusion in the `/contrib` directory)? Just thinking out loud

moxiegirl commented:
@chenchun Thanks for the incorporation. LGTM for docs.  @thaJeztah I tested mbentley's add yesterday.  After this merges, I have an update there.

chenchun commented:
Just updated the PR to address @rhvgoyal 's comments.

thaJeztah commented:
LGTM

ping @moxiegirl for the merge

moxiegirl commented:
@chenchun Thanks for the quick turnaround!
